File: Graph/TopologicalSort.py
#verify: https://onlinejudge.u-aizu.ac.jp/courses/library/5/GRL/all/GRL_4_B

res = []
visited = []
def dfs(n,adj):
    global visited
    visited[n] = 1
    for i in adj[n]:
        if not visited[i]:dfs(i,adj)
    res.append(n)
 
def tsort(n,adj):
    for i in range(n):
        if not visited[i]:dfs(i,adj)
    res.reverse()
    return res
 
v,e = map(int,input().split())
visited = [0] * v
adj = [[] * v for _ in range(v)]
for _ in range(e):
    s,t = map(int,input().split())
    adj[s].append(t)
 
res = tsort(v,adj)
for i in res:print(i)

File: C++/GraphTheory/TopologicalSort.cpp
vector<int>tsort(vector<vector<int>> v) {
	int n = v.size();
	vector<int>mon(n, 0), res;
	stack<int>s;
	for (int i = 0; i < n; ++i) {
		for (auto j : v[i]) { ++mon[j]; }
	}
	for (int i = 0; i < n; ++i) {
		if (mon[i] == 0) { s.push(i); }
	}
	while (s.size()) {
		int mmon = s.top();
		s.pop();
		res.push_back(mmon);
		for (auto i : v[mmon]) {
			--mon[i];
			if (mon[i] == 0) { s.push(i); }
		}
	}
	return res;
}

/*verify: https://onlinejudge.u-aizu.ac.jp/problems/GRL_4_B
int main() {
	ll v, e;
	cin >> v >> e;
	vector<vector<int>>gr(v);
	for (int i = 0; i < e; ++i) {
		int s, t;
		cin >> s >> t;
		gr[s].push_back(t);
	}
	auto res = tsort(gr);
	for (auto i : res) { cout << i << endl; }
}
*/